File: tests/python/extensions/pybind/cuda/core/device.py
#!/usr/bin/env python3

import numpy.testing as npt
import pybind_isce3 as isce3

def test_get_device_count():
    count = isce3.cuda.core.get_device_count()
    assert(count >= 0)

def test_init():
    count = isce3.cuda.core.get_device_count()
    for d in range(count):
        device = isce3.cuda.core.Device(d)
        assert(device.id == d)

        print("Device %i" % (device.id))
        print("--------")
        print("name: %s" % (device.name))
        print("compute: %s" % (device.compute_capability))
        print("total mem (bytes): %i" % (device.total_global_mem))

        assert(device.name != "")
        assert(device.total_global_mem > 0)
        assert(device.compute_capability.major >= 1)
        assert(device.compute_capability.minor >= 0)

def test_invalid_device():
    with npt.assert_raises(ValueError):
        device = isce3.cuda.core.Device(-1)

    count = isce3.cuda.core.get_device_count()
    with npt.assert_raises(ValueError):
        device = isce3.cuda.core.Device(count)

def test_get_device():
    device = isce3.cuda.core.get_device()
    assert(device.id >= 0)

def test_set_device():
    count = isce3.cuda.core.get_device_count()
    for d in range(count):
        device = isce3.cuda.core.Device(d)
        isce3.cuda.core.set_device(d)
        assert(isce3.cuda.core.get_device().id == d)

def test_comparison():
    device1 = isce3.cuda.core.Device(0)
    device2 = isce3.cuda.core.Device(0)
    assert(device1 == device2)

    count = isce3.cuda.core.get_device_count()
    if (count > 1):
        device3 = isce3.cuda.core.Device(1)
        assert(device1 != device3)

File: tests/cxx/isce3/cuda/core/device.cpp
#include <gtest/gtest.h>

#include <isce3/cuda/core/Device.h>
#include <isce3/except/Error.h>

using isce3::cuda::core::Device;
using isce3::cuda::core::getDevice;
using isce3::cuda::core::getDeviceCount;
using isce3::cuda::core::setDevice;
using isce3::except::InvalidArgument;

struct DeviceTest : public testing::Test {

    void SetUp() override
    {
        count = getDeviceCount();
        EXPECT_GE(count, 1);
    }

    int count = 0;
};

TEST_F(DeviceTest, Device)
{
    for (int id = 0; id < count; ++id) {
        const Device device(id);
        EXPECT_EQ(device.id(), id);

        std::cout << "Device " << device.id() << std::endl
                  << "--------" << std::endl
                  << "name: " << device.name() << std::endl
                  << "compute: " << device.computeCapability() << std::endl
                  << "total mem (bytes): " << device.totalGlobalMem() << std::endl;

        EXPECT_NE(device.name(), "");
        EXPECT_GT(device.totalGlobalMem(), 0);

        const auto compute = device.computeCapability();
        EXPECT_GE(compute.major, 1);
        EXPECT_GE(compute.minor, 0);
    }
}

TEST_F(DeviceTest, InvalidDevice)
{
    EXPECT_THROW({ const Device device(-1); }, InvalidArgument);
    EXPECT_THROW({ const Device device(count); }, InvalidArgument);
}

TEST_F(DeviceTest, GetDevice)
{
    const Device device = getDevice();
    EXPECT_GE(device.id(), 0);
}

TEST_F(DeviceTest, SetDevice)
{
    for (int id = 0; id < count; ++id) {
        const Device device(id);
        setDevice(id);
        EXPECT_EQ(getDevice(), device);
    }
}

TEST_F(DeviceTest, Comparison)
{
    const Device device1(0);
    const Device device2(0);
    EXPECT_TRUE(device1 == device2);

    if (count > 1) {
        const Device device3(1);
        EXPECT_TRUE(device1 != device3);
    }
}

int main(int argc, char** argv)
{
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
